precompile/allow_list.go

```diff
@@ -147,49 +146,55 @@ func (mal *modifyAllowListPrecompile) Run(evm PrecompileAccessibleState, callerA
 
 	// Verify that the caller is in the allow list and therefore has the right to modify it
 	callerStatus := GetAllowListStatus(evm.GetStateDB(), callerAddr)
-	if !callerStatus.IsAdmin() {
-		log.Info("EVM received attempt to modify the allow list from a non-allowed address", "callerAddr", callerAddr)
+	if !IsAllowListAdmin(callerStatus) {
+		log.Info("allow list EVM received attempt to modify the allow list from a non-allowed address", "callerAddr", callerAddr, "callerStatus", callerStatus)
 		return nil, remainingGas, fmt.Errorf("caller %s cannot modify allow list", callerAddr)
 	}
 
-	// Unpack the precompile's input into the arguments for setAllowListStatus
-	address, status, err := UnpackModifyAllowList(input)
-	if err != nil {
-		log.Info("modify allow list reverted", "err", err)
-		return nil, remainingGas, fmt.Errorf("failed to unpack modify allow list input: %w", err)
-	}
-
-	setAllowListStatus(evm.GetStateDB(), address, status)
+	SetAllowListRole(evm.GetStateDB(), modifyAddress, role)
 
 	// Return an empty output and the remaining gas
 	return []byte{}, remainingGas, nil
 }
 
-// RequiredGas returns the amount of gas consumed by this precompile.
-func (mal *modifyAllowListPrecompile) RequiredGas(input []byte) uint64 { return ModifyAllowListGasCost }
+// createAllowListSetter returns an execution function for setting the allow list status of the input address argument to [role].
+func createAllowListSetter(role common.Hash) func(evm PrecompileAccessibleState, callerAddr common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	return func(evm PrecompileAccessibleState, callerAddr, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+		if len(input) != common.HashLength {
+			return nil, remainingGas, fmt.Errorf("invalid input length for modifying allow list: %d", len(input))
+		}
 
-// readAllowListPrecompile implements StatefulPrecompiledContract and can be used as a thread-safe singleton.
-// Provides read access to the contract deployer allow list.
-type readAllowListPrecompile struct{}
+		modifyAddress := common.BytesToAddress(input)
+		return writeAllowList(evm, callerAddr, modifyAddress, role, suppliedGas, readOnly)
+	}
+}
 
-// Run implements StatefulPrecompiledContract
-// parses [input] into a single address and returns the 32 byte hash that specifies the designated role of that address.
-func (ral *readAllowListPrecompile) Run(evm PrecompileAccessibleState, callerAddr common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+// readAllowList parses [input] into a single address and returns the 32 byte hash that specifies the designated role of that address.
```

Comment:
nit: add comment explaining how this is used to implement a given function signature.